Bluetooth: audio: BAP Broadcast Assistant support for multiple connections

pin-zephyr opened the pull request.

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -879,51 +939,83 @@ static int broadcast_assistant_reset(struct bap_broadcast_assistant_instance *in
-	(void)broadcast_assistant_reset(&broadcast_assistant);
+	struct bap_broadcast_assistant_instance *inst = inst_by_conn(conn);
+
+	if (!inst) {
+		return;
+	}
+
+	(void)broadcast_assistant_reset(inst);
```
```suggestion
	if (inst) {
		(void)broadcast_assistant_reset(inst);
	}
```

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -879,51 +939,83 @@ static int broadcast_assistant_reset(struct bap_broadcast_assistant_instance *in
+	(void)broadcast_assistant_reset(inst);
+}
+
+static void connected_cb(struct bt_conn *conn, uint8_t err)
+{
+	struct bap_broadcast_assistant_instance *inst;
+
+	if (err != 0U) {
+		LOG_DBG("Failed to connect to %p (%u)", conn, err);
+		return;
+	}
+
+	LOG_DBG("Connected: %p", conn);
+
+	inst = &broadcast_assistants[bt_conn_index(conn)];
+
+	if (inst->conn == conn) {
+		return;
+	}
+
+	if (broadcast_assistant_reset(inst) != 0) {
+		LOG_DBG("Failed to reset broadcast assistant: %d", err);
+
+		return;
+	}
+
+	inst->conn = bt_conn_ref(conn);
+	inst->busy = false;
```
What is the purpose of this? 

We do not support bonding yet, and even if we did, we cannot rely on the conn index being the same. 

The `if (inst->conn == conn) {` check should always fail as clear the `conn` pointer in the disconnected. 

I don't think we have any use for this function and it should be removed again.

pin-zephyr replied:
here I try to bind a struct bt_conn *conn with a specific bap_broadcast_assistant_instance via the unique conn index.  Since it is not feasible to use CONTAINER_OF() for struct bt_conn *conn to retrieve bap_broadcast_assistant_instance later on.

do you have a better suggesion to bind a struct bt_conn *conn with bap_broadcast_assistant_instance ?

Thalley replied:
> do you have a better suggesion to bind a struct bt_conn *conn with bap_broadcast_assistant_instance ?

The `conn` is bound to the assistant when calling `bt_bap_broadcast_assistant_discover`

Thalley reviewed (changes requested):
Excellent work

One comment about the `connected_cb`

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -280,6 +292,11 @@ static uint8_t parse_and_send_recv_state(struct bt_conn *conn, uint16_t handle,
 	int16_t index;
+	struct bap_broadcast_assistant_instance *inst = inst_by_conn(conn);
+
+	if (!inst) {
```
That is also my preference. The only reason why I didn't request this as a change, is that it's not required by our coding guidelines :)

pin-zephyr replied:
I see the advantage, it will nice to mention in the coding guidelines.

Thalley replied:
It is technically there as MISRA rule 14.4 (https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#main-rules), but we are not yet at a stage in the coding guidelines where we (as collaborators) are "allowed" to reject PRs not conforming to the MISRA rules.

pin-zephyr commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -108,6 +109,17 @@ static int16_t lookup_index_by_handle(uint16_t handle)
 
+static struct bap_broadcast_assistant_instance *inst_by_conn(struct bt_conn *conn)
+{
+	struct bap_broadcast_assistant_instance *inst = &broadcast_assistants[bt_conn_index(conn)];
```
it has already been guard before calling inst_by_conn() in the cases which conn could be NULL

Thalley replied:
As long as this is a `static` function I'm OK with not having the `conn == NULL` check, as we can ensure that where we call this. 

That being said, `notify_handler` does call this without a check, and IIRC that function can be called from GATT with `conn == NULL` if there is a disconnect on an unbonded device

It may be better to simply add the check in this function to make sure

Thalley reviewed (changes requested):
My suggestion from https://github.com/zephyrproject-rtos/zephyr/pull/73694/files#r1631213822 should be implemented, but otherwise LGTM

pin-zephyr committed:
Bluetooth: audio: BAP Broadcast Assistant support for multiple connections

The current implementation of the BAP Broadcast Assistant only supports a
single connection. This PR makes broadcast assistant support multiple BAP
scan delegators.

Fixes: #67523

Signed-off-by: Ping Wang <pinw@demant.com>

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -108,6 +109,17 @@ static int16_t lookup_index_by_handle(uint16_t handle)
+{
+	struct bap_broadcast_assistant_instance *inst = &broadcast_assistants[bt_conn_index(conn)];
+
+	if (conn == NULL || inst->conn != conn) {
+		return NULL;
+	}
```
The `conn` check needs to be before the call to `bt_conn_index`

pin-zephyr replied:
just for my understanding, why can not put the check inside of the function, it can avoid copy the check code in multiple places ?

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -108,6 +109,24 @@ static int16_t lookup_index_by_handle(uint16_t handle)
+	struct bap_broadcast_assistant_instance *inst;
+
+	if (conn == NULL) {
+		LOG_ERR("NULL conn");
```
```suggestion
		LOG_DBG("NULL conn");
```

We should not have an error here, as this can happen in non-error cases (which is why we needed the check).
